### PR TITLE
ARROW-4648: [C++] Use underscores in source file names

### PR DESCRIPTION
**Pull request by pitrou**
The rule is:
- underscores in C++ source file names
- underscores in .so / .a file names
- hyphens in executable file names
- hyphens in pkgconfig file names

### CODE REVIEW

**Comment by emkornfield** [comment]
I perused this, and it looks good to me.  It seems that this change triggered a bunch of IWYU lint issues.

**Comment by pitrou** [comment]
@emkornfield Yes, I'll fix the lint issues at the end.

File: cpp/build-support/lint_cpp_cli.py [hunk 1]
```diff
@@ -103,6 +103,13 @@ def lint_files():
             if exclude:
                 continue
 
+            # Lint file name, except for pkgconfig templates
+            if not filename.endswith('.pc.in'):
+                if '-' in filename:
+                    why = ("Please user underscores, not hyphens, "
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/build-support/lint_cpp_cli.py]: Thanks, will fix :-)

**Comment by wesm** [comment]
Looks good to me also.

**Comment by pitrou** [comment]
Rebased and hopefully fixed lint failures. Will wait for CI and merge if green.

**Comment by pitrou** [comment]
Travis-CI: https://travis-ci.org/pitrou/arrow/builds/573853109
There's a hung Travis-CI job. It also happens on master, so not related to this PR: https://travis-ci.org/apache/arrow/builds/573842140

**Comment by pitrou** [comment]
AppVeyor: https://ci.appveyor.com/project/pitrou/arrow/builds/26798788

**Pull request closed without merging**